File: constantine/ethereum_verkle_ipa.nim
# Constantine
# Copyright (c) 2018-2019    Status Research & Development GmbH
# Copyright (c) 2020-Present Mamy André-Ratsimbazafy
# Licensed and distributed under either of
#   * MIT license (license terms in the root directory or at http://opensource.org/licenses/MIT).
#   * Apache v2 license (license terms in the root directory or at http://www.apache.org/licenses/LICENSE-2.0).
# at your option. This file may not be copied, modified, or distributed except according to those terms.

const prefix_ffi = "ctt_eth_verkle_ipa_"

import ./zoo_exports

import
  ./platforms/[abstractions, views],
  ./named/zoo_subgroups,
  ./commitments/eth_verkle_ipa,
  ./math/[arithmetic, ec_twistededwards, extension_fields],
  constantine/named/algebras,
  ./serialization/[codecs_status_codes, codecs_banderwagon],
  constantine/math/polynomials/polynomials,
  ./commitments/eth_verkle_transcripts

# ------------------------------------------------------------
# TODO: finish refactoring generate_random_points
import
  ./platforms/allocs,
  ./hashes,
  ./serialization/endians,
  ./math/io/[io_bigints, io_fields]

export 
  hashes,
  abstractions,
  ec_twistededwards,
  CttCodecScalarStatus,
  CttCodecEccStatus

{.checks: off.}

const EthVerkleSeed* = "eth_verkle_oct_2021"
export EthVerkleTranscript 
  
# func add(a: int, b: int): int {.libPrefix: prefix_ffi.} =
#     return a + b
func generate_random_points*(r: var openArray[EC_TwEdw_Aff[Fp[Banderwagon]]]) =
  ## generate_random_points generates random points on the curve with the hardcoded EthVerkleSeed
  let points = allocHeapArrayAligned(EC_TwEdw_Aff[Fp[Banderwagon]], r.len, alignment = 64)

  var points_found: seq[EC_TwEdw_Aff[Fp[Banderwagon]]]
  var incrementer: uint64 = 0
  var idx: int = 0
  while true:
    var ctx {.noInit.}: sha256
    ctx.init()
    ctx.update(EthVerkleSeed)
    ctx.update(incrementer.toBytes(bigEndian))
    var hash : array[32, byte]
    ctx.finish(hash)
    ctx.clear()

    var x {.noInit.}:  Fp[Banderwagon]
    var t {.noInit.}: Fp[Banderwagon].getBigInt()

    t.unmarshal(hash, bigEndian)
    x.fromBig(t)

    incrementer = incrementer + 1

    var x_arr {.noInit.}: array[32, byte]
    x_arr.marshal(x, bigEndian)

    var x_p {.noInit.}: EC_TwEdw_Aff[Fp[Banderwagon]]
    let stat2 = x_p.deserialize_vartime(x_arr)
    if stat2 == cttCodecEcc_Success:
      points_found.add(x_p)
      points[idx] = points_found[idx]
      idx = idx + 1

    if points_found.len == r.len:
      break

  for i in 0 ..< r.len:
    r[i] = points[i]
  freeHeapAligned(points)

# ------------------------------------------------------------

# Ethereum Verkle IPA public API
# ------------------------------------------------------------
#
# We use a simple goto state machine to handle errors and cleanup (if allocs were done)
# and have 2 different checks:
# - Either we are in "HappyPath" section that shortcuts to resource cleanup on error
# - or there are no resources to clean and we can early return from a function.

const EthVerkleDomain* = 256

type
  cttEthVerkleIpaStatus* = enum
    cttEthVerkleIpa_Success
    cttEthVerkleIpa_VerificationFailure
    cttEthVerkleIpa_InputsLengthsMismatch
    cttEthVerkleIpa_ScalarZero
    cttEthVerkleIpa_ScalarLargerThanCurveOrder
    cttEthVerkleIpa_EccInvalidEncoding
    cttEthVerkleIpa_EccCoordinateGreaterThanOrEqualModulus
    cttEthVerkleIpa_EccPointNotOnCurve
    cttEthVerkleIpa_EccPointNotInSubGroup

template checkReturn(evalExpr: CttCodecScalarStatus): untyped {.dirty.} =
  # Translate codec status code to KZG status code
  # Beware of resource cleanup like heap allocation, this can early exit the caller.
  block:
    let status = evalExpr # Ensure single evaluation
    case status
    of cttCodecScalar_Success:                          discard
    of cttCodecScalar_Zero:                             discard
    of cttCodecScalar_ScalarLargerThanCurveOrder:       return cttEthVerkleIpa_ScalarLargerThanCurveOrder

template checkReturn(evalExpr: CttCodecEccStatus): untyped {.dirty.} =
  # Translate codec status code to KZG status code
  # Beware of resource cleanup like heap allocation, this can early exit the caller.
  block:
    let status = evalExpr # Ensure single evaluation
    case status
    of cttCodecEcc_Success:                             discard
    of cttCodecEcc_InvalidEncoding:                     return cttEthVerkleIpa_EccInvalidEncoding
    of cttCodecEcc_CoordinateGreaterThanOrEqualModulus: return cttEthVerkleIpa_EccCoordinateGreaterThanOrEqualModulus
    of cttCodecEcc_PointNotOnCurve:                     return cttEthVerkleIpa_EccPointNotOnCurve
    of cttCodecEcc_PointNotInSubgroup:                  return cttEthVerkleIpa_EccPointNotInSubGroup
    of cttCodecEcc_PointAtInfinity:                     discard

template check(Section: untyped, evalExpr: CttCodecScalarStatus): untyped {.dirty.} =
  # Translate codec status code to KZG status code
  # Exit current code block
  block:
    let status = evalExpr # Ensure single evaluation
    case status
    of cttCodecScalar_Success:                          discard
    of cttCodecScalar_Zero:                             discard
    of cttCodecScalar_ScalarLargerThanCurveOrder:       result = cttEthVerkleIpa_ScalarLargerThanCurveOrder; break Section

template check(Section: untyped, evalExpr: CttCodecEccStatus): untyped {.dirty.} =
  # Translate codec status code to KZG status code
  # Exit current code block
  block:
    let status = evalExpr # Ensure single evaluation
    case status
    of cttCodecEcc_Success:                             discard
    of cttCodecEcc_InvalidEncoding:                     result = cttEthVerkleIpa_EccInvalidEncoding; break Section
    of cttCodecEcc_CoordinateGreaterThanOrEqualModulus: result = cttEthVerkleIpa_EccCoordinateGreaterThanOrEqualModulus; break Section
    of cttCodecEcc_PointNotOnCurve:                     result = cttEthVerkleIpa_EccPointNotOnCurve; break Section
    of cttCodecEcc_PointNotInSubgroup:                  result = cttEthVerkleIpa_EccPointNotInSubGroup; break Section
    of cttCodecEcc_PointAtInfinity:                     discard

# Serialization
# ------------------------------------------------------------------------------------

type
  EthVerkleIpaProofBytes* {.exportc: prefix_ffi & "proof_bytes".}= array[544, byte]
  EthVerkleIpaMultiProofBytes* {.exportc: prefix_ffi & "multi_proof_bytes".}= array[576, byte]
  EthVerkleIpaProof* {.exportc: prefix_ffi & "proof".} = object
    aff: IpaProof[8, EC_TwEdw_Aff[Fp[Banderwagon]], Fr[Banderwagon]]
    prj: IpaProof[8, EC_TwEdw_Prj[Fp[Banderwagon]], Fr[Banderwagon]]
  EthVerkleIpaMultiProof* {.exportc: prefix_ffi & "multi_proof".} = object
    aff: IpaMultiProof[8, EC_TwEdw_Aff[Fp[Banderwagon]], Fr[Banderwagon]]
    prj: IpaMultiProof[8, EC_TwEdw_Prj[Fp[Banderwagon]], Fr[Banderwagon]]
  EthVerkleIpaPolyEvalCrs* {.exportc: prefix_ffi & "poly_eval".} = object
    aff: PolynomialEval[EthVerkleDomain, EC_TwEdw_Aff[Fp[Banderwagon]]]
    prj: PolynomialEval[EthVerkleDomain, EC_TwEdw_Prj[Fp[Banderwagon]]]
  EthVerkleIpaPolyEvalPoly* {.exportc: prefix_ffi & "poly_eval_poly".} = object
    raw: PolynomialEval[256, Fr[Banderwagon]]

  # The aliases may throw strange errors like:
  # - Error: invalid type: 'EthVerkleIpaProof' for var
  # - Error: cannot instantiate: 'src:type'
  # as of Nim v2.0.4

func serialize*(dst: var EthVerkleIpaProofBytes,
                src: IpaProof[8, EC_TwEdw[Fp[Banderwagon]], Fr[Banderwagon]]
                ): cttEthVerkleIpaStatus {.discardable.} =
  # Note: We store 1 out of 2 coordinates of an EC point, so size(Fp[Banderwagon])
  const fpb = sizeof(Fp[Banderwagon])
  const frb = sizeof(Fr[Banderwagon])

  let L = cast[ptr array[8, array[fpb, byte]]](dst.addr)
  let R = cast[ptr array[8, array[fpb, byte]]](dst[8 * fpb].addr)
  let a0 = cast[ptr array[frb, byte]](dst[2 * 8 * fpb].addr)

  for i in 0 ..< 8:
    L[i].serialize(src.L[i])

  for i in 0 ..< 8:
    R[i].serialize(src.R[i])

  a0[].serialize_fr(src.a0, littleEndian)
  return cttEthVerkleIpa_Success

func serialize_aff*(dst: var EthVerkleIpaProofBytes,
                src: EthVerkleIpaProof
                ): cttEthVerkleIpaStatus {.libPrefix: prefix_ffi.} =
  
  return serialize(dst, src.aff)

func serialize_prj*(dst: var EthVerkleIpaProofBytes,
                src: EthVerkleIpaProof
                ): cttEthVerkleIpaStatus {.libPrefix: prefix_ffi.} =
  
  return serialize(dst, src.prj)

func deserialize*(dst: var IpaProof[8, EC_TwEdw[Fp[Banderwagon]], Fr[Banderwagon]],
                  src: EthVerkleIpaProofBytes): cttEthVerkleIpaStatus {.discardable.} =

  const fpb = sizeof(Fp[Banderwagon])
  const frb = sizeof(Fr[Banderwagon])

  let L = cast[ptr array[8, array[fpb, byte]]](src.unsafeAddr)
  let R = cast[ptr array[8, array[fpb, byte]]](src[8 * fpb].unsafeAddr)
  let a0 = cast[ptr array[frb, byte]](src[2 * 8 * fpb].unsafeAddr)

  for i in 0 ..< 8:
    checkReturn dst.L[i].deserialize_vartime(L[i])

  for i in 0 ..< 8:
    checkReturn dst.R[i].deserialize_vartime(R[i])

  checkReturn dst.a0.deserialize_fr(a0[], littleEndian)
  return cttEthVerkleIpa_Success

func deserialize_aff*(dst: var EthVerkleIpaProof,
                  src: EthVerkleIpaProofBytes): cttEthVerkleIpaStatus {.libPrefix:prefix_ffi.} =
  return deserialize(dst.aff, src)

# func deserialize_prj*(dst: var EthVerkleIpaProof,
#                   src: EthVerkleIpaProofBytes): cttEthVerkleIpaStatus {.libPrefix: prefix_ffi.} =
#   return deserialize(dst.prj, src)

func batch_serialize*(dst: var EthVerkleIpaMultiProofBytes,
                src: IpaMultiProof[8, EC_TwEdw[Fp[Banderwagon]], Fr[Banderwagon]]
                ): cttEthVerkleIpaStatus {.discardable.} =

  const frb = sizeof(Fr[Banderwagon])
  let D = cast[ptr array[frb, byte]](dst.addr)
  let g2Proof = cast[ptr EthVerkleIpaProofBytes](dst[frb].addr)

  D[].serialize(src.D)
  g2Proof[].serialize(src.g2_proof)
  return cttEthVerkleIpa_Success

func batch_serialize_aff*(dst: var EthVerkleIpaMultiProofBytes,
                src: EthVerkleIpaMultiProof
                ): cttEthVerkleIpaStatus {.libPrefix: prefix_ffi.} =
  
  return batch_serialize(dst, src.aff)

func batch_serialize_prj*(dst: var EthVerkleIpaMultiProofBytes,
                src: EthVerkleIpaMultiProof
                ): cttEthVerkleIpaStatus {.libPrefix: prefix_ffi.} =
  
  return batch_serialize(dst, src.prj)

func batch_deserialize*(dst: var IpaMultiProof[8, EC_TwEdw[Fp[Banderwagon]], Fr[Banderwagon]],
                  src: EthVerkleIpaMultiProofBytes
                  ): cttEthVerkleIpaStatus =

  const frb = sizeof(Fr[Banderwagon])
  let D = cast[ptr array[frb, byte]](src.unsafeAddr)
  let g2Proof = cast[ptr EthVerkleIpaProofBytes](src[frb].unsafeAddr)

  checkReturn dst.D.deserialize_vartime(D[])
  return dst.g2_proof.deserialize(g2Proof[])

func batch_deserialize_aff*(dst: var EthVerkleIpaMultiProof,
                  src: EthVerkleIpaMultiProofBytes
                  ): cttEthVerkleIpaStatus {.libPrefix: prefix_ffi.} =
  
  return batch_deserialize(dst.aff, src)

# Mapping EC to scalars
# ------------------------------------------------------------------------------------

# TODO: refactor, this shouldn't use curves_primitives but internal functions
import ./lowlevel_fields

func map_to_base_field*(dst: var Fp[Banderwagon],p: EC_TwEdw[Fp[Banderwagon]]) {.discardable.} =
  ## The mapping chosen for the Banderwagon Curve is x/y
  ##
  ## This function takes a Banderwagon element & then
  ## computes the x/y value and returns as an Fp element
  ##
  ## Spec : https://hackmd.io/@6iQDuIePQjyYBqDChYw_jg/BJBNcv9fq#Map-To-Field

  var invY: Fp[Banderwagon]
  invY.inv(p.y)             # invY = 1/Y
  dst.prod(p.x, invY)       # dst = (X) * (1/Y)

func map_to_base_field_aff*(dst: var Fp[Banderwagon], p: EC_TwEdw_Aff[Fp[Banderwagon]]) {.libPrefix: prefix_ffi.} =
  map_to_base_field(dst, p)

func map_to_base_field_prj*(dst: var Fp[Banderwagon], p: EC_TwEdw_Prj[Fp[Banderwagon]]) {.libPrefix: prefix_ffi.} =
  map_to_base_field(dst, p)

func map_to_scalar_field*(res: var Fr[Banderwagon], p: EC_TwEdw[Fp[Banderwagon]]): bool {.discardable.} =
  ## This function takes the x/y value from the above function as Fp element
  ## and convert that to bytes in Big Endian,
  ## and then load that to a Fr element
  ##
  ## Spec : https://hackmd.io/wliPP_RMT4emsucVuCqfHA?view#MapToFieldElement

  var baseField: Fp[Banderwagon]
  var baseFieldBytes: array[32, byte]

  baseField.mapToBaseField(p)   # compute the defined mapping

  let check1 = baseFieldBytes.marshalBE(baseField)  # Fp -> bytes
  let check2 = res.unmarshalBE(baseFieldBytes)      # bytes -> Fr

  return check1 and check2

func map_to_scalar_field_aff*(res: var Fr[Banderwagon], p: EC_TwEdw_Aff[Fp[Banderwagon]]): bool {.libPrefix: prefix_ffi.} =
  return map_to_scalar_field(res, p)

func map_to_scalar_field_prj*(res: var Fr[Banderwagon], p: EC_TwEdw_Prj[Fp[Banderwagon]]): bool {.libPrefix: prefix_ffi.} =
  return map_to_scalar_field(res, p)

func batch_map_to_scalar_field*(
      res: var openArray[Fr[Banderwagon]],
      points: openArray[EC_TwEdw[Fp[Banderwagon]]]): bool {.discardable, noinline.} =
  ## This function performs the `mapToScalarField` operation
  ## on a batch of points
  ##
  ## The batch inversion used in this using
  ## the montogomenry trick, makes is faster than
  ## just iterating of over the array of points and
  ## converting the curve points to field elements
  ##
  ## Spec : https://hackmd.io/wliPP_RMT4emsucVuCqfHA?view#MapToFieldElement

  var check: bool = true
  check = check and (res.len == points.len)

  let N = res.len
  var ys = allocStackArray(Fp[Banderwagon], N)
  var ys_inv = allocStackArray(Fp[Banderwagon], N)


  for i in 0 ..< N:
    ys[i] = points[i].y

  ys_inv.batchInv_vartime(ys, N)

  for i in 0 ..< N:
    var mappedElement: Fp[Banderwagon]
    var bytes: array[32, byte]

    mappedElement.prod(points[i].x, ys_inv[i])
    check = bytes.marshalBE(mappedElement)
    check = check and res[i].unmarshalBE(bytes)

  return check

func batch_map_to_scalar_field_aff*(
      res: var openArray[Fr[Banderwagon]],
      points: openArray[EC_TwEdw_Aff[Fp[Banderwagon]]]): bool {.libPrefix: prefix_ffi.} =
  return batch_map_to_scalar_field(res, points)

func batch_map_to_scalar_field_prj*(
      res: var openArray[Fr[Banderwagon]],
      points: openArray[EC_TwEdw_Prj[Fp[Banderwagon]]]): bool {.libPrefix: prefix_ffi.} =
  return batch_map_to_scalar_field(res, points)
# Inner Product Argument
# ------------------------------------------------------------------------------------
# TODO: proper IPA wrapper for https://github.com/status-im/nim-eth-verkle
#
# For now we reexport
# - eth_verkle_ipa
# - sha256 for transcripts

func commit*(
      crs: EthVerkleIpaPolyEvalCrs,
      r: var EC_TwEdw_Prj[Fp[Banderwagon]], 
      poly: EthVerkleIpaPolyEvalPoly) {.libPrefix: prefix_ffi.} =
  ipa_commit(crs.aff, r, poly.raw)

func prove*(
      crs: EthVerkleIpaPolyEvalCrs, 
      domain: PolyEvalLinearDomain[EthVerkleDomain, Fr[Banderwagon]],
      transcript_label: string,
      eval_at_challenge: var Fr[Banderwagon], 
      proof: var IpaProof[8, EC_TwEdw_Aff[Fp[Banderwagon]], Fr[Banderwagon]], 
      poly: EthVerkleIpaPolyEvalPoly, 
      commitment: EC_TwEdw_Aff[Fp[Banderwagon]], 
      opening_challenge: Fr[Banderwagon]) {.libPrefix: prefix_ffi.} =
  var transcript: sha256
  transcript.initTranscript(transcript_label)
  ipa_prove(crs.aff, domain, transcript, eval_at_challenge, proof, poly.raw, commitment, opening_challenge)

func verify*(
      crs: EthVerkleIpaPolyEvalCrs, 
      domain: PolyEvalLinearDomain[EthVerkleDomain, Fr[Banderwagon]], 
      transcript_label: string, 
      commitment: EC_TwEdw_Aff[Fp[Banderwagon]],
      opening_challenge: Fr[Banderwagon],
      eval_at_challenge: var Fr[Banderwagon], 
      proof: IpaProof[8, EC_TwEdw_Aff[Fp[Banderwagon]], Fr[Banderwagon]]): bool {.libPrefix: prefix_ffi.} =
  var transcript: sha256
  transcript.initTranscript(transcript_label)
  return ipa_verify(crs.aff, domain, transcript, commitment, opening_challenge, eval_at_challenge, proof)

func multi_prove*(
      crs: EthVerkleIpaPolyEvalCrs, 
      domain: PolyEvalLinearDomain[EthVerkleDomain, Fr[Banderwagon]], 
      transcript_label: string, 
      proof: var IpaMultiProof[8, EC_TwEdw_Aff[Fp[Banderwagon]], Fr[Banderwagon]], 
      polys: openArray[PolynomialEval[256, Fr[Banderwagon]]], 
      commitments: openArray[EC_TwEdw_Aff[Fp[Banderwagon]]], 
      opening_challenges_in_domain: openArray[uint64]) {.libPrefix: prefix_ffi.} =
  var transcript: sha256
  transcript.initTranscript(transcript_label)
  ipa_multi_prove(crs.aff, domain, transcript, proof, polys, commitments, opening_challenges_in_domain)

func multi_verify*(
      crs: EthVerkleIpaPolyEvalCrs, 
      domain: PolyEvalLinearDomain[EthVerkleDomain, Fr[Banderwagon]], 
      transcript_label: string, 
      commitments: openArray[EC_TwEdw_Aff[Fp[Banderwagon]]],
      opening_challenges_in_domain: openArray[uint64],
      evals_at_challenges: openArray[Fr[Banderwagon]], 
      proof: IpaMultiProof[8, EC_TwEdw_Aff[Fp[Banderwagon]], Fr[Banderwagon]]): bool {.libPrefix: prefix_ffi.} =
  var transcript: sha256
  transcript.initTranscript(transcript_label)
  return ipa_multi_verify(crs.aff, domain, transcript, commitments, opening_challenges_in_domain, evals_at_challenges, proof)

export eth_verkle_ipa

File: examples-c/ethereum_verkle_ipa.c
/** Constantine
 *  Copyright (c) 2018-2019    Status Research & Development GmbH
 *  Copyright (c) 2020-Present Mamy André-Ratsimbazafy
 *  Licensed and distributed under either of
 *    * MIT license (license terms in the root directory or at http://opensource.org/licenses/MIT).
 *    * Apache v2 license (license terms in the root directory or at http://www.apache.org/licenses/LICENSE-2.0).
 *  at your option. This file may not be copied, modified, or distributed except according to those terms.
 */

#include <assert.h>
#include <stdio.h>
#include <stdlib.h>

#include <string.h>
#include <constantine.h>

void hex_to_bytes(const char *hex_string, byte *byte_array, size_t array_size) {
    size_t hex_len = strlen(hex_string);
    if (hex_len % 2 != 0 || hex_len / 2 > array_size) {
        fprintf(stderr, "Error: Hex string length invalid or too large.\n");
        return;
    }
    for (size_t i = 0; i < hex_len / 2; i++) {
        sscanf(&hex_string[2 * i], "%2hhx", &byte_array[i]);
    }
}

// TODO: complete testing
int main(){
  // Protocol and deserialization statuses
  ctt_eth_verkle_ipa_status      ipa_status;

  banderwagon_ec_prj point;
  char *endptr;
  char *endptrX;
  char *endptrY;
  char *endptrZ;
  // point.x = (banderwagon_fp) { .limbs = { strtol("0x0e7e3748db7c5c999a7bcd93d71d671f1f40090423792266f94cb27ca43fce5c", &endptrX, 16) } };
  // point.y = (banderwagon_fp) { .limbs = { strtol("0x563a625521456130dc66f9fd6bda67330c7bb183b7f2223216c1c9536e1c622f", &endptrY, 16) } };
  // point.z = (banderwagon_fp) { .limbs = { strtol("0x01", &endptrZ, 16) } };

  const ctt_eth_verkle_ipa_proof_bytes proof_bytes = { {0} };

  hex_to_bytes("00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002d3e383cf2ca36482707617daf4230f2261cff2abeb98a7d1e139cf386970f7a67cea4e0dcf8c437e5cd9852d95613a255ef625412a3ac7fb1a0d27227a32a7c1292f14b7c189f033c91217f02b34c7832958afc7ae3bb498b29ca08277dc60d1c53bb5f07280c16238a7f99c059cbbdbbc933bef4b74d604721a09b526aac1751a4bdf0df2d303418e7e5642ac4aacc730625514c87a4bcce5369cc4c1e1d2a1ee9125e09db763e7d99fa857928fabeb94ba822d5cf1cc8f5be372683ee7089082c0ca302a243f0124cc25319d069e0c689f03e4cb32e266fffd4b8c9a5e1cb2c708dc7960531ecea4331e376d7f6604228fc0606a08bda95ee3350c8bca83f37b23160af7bae3db95f0c66ed4535fc5397b43dcdc1d09c1e3a0376a6705d916d96cb64feb47d00ebf1ddbad7eaf3b5d8c381d31098c5c8a909793bd6063c2f0450320af78de387938261eba3e984271f31c3f71a55b33631b90505f8209b384aa55feb1c1c72a5e2abce15f24eb18715a309f5517ac3079c64c8ff157d3e35d5bad17b86f9599b1e34f1f4b7c6600a83913261645a0811fba0ad1ed104fe0c", proof_bytes.raw, sizeof(proof_bytes.raw));
  // printf("Converted raw bytes (%zu bytes):\n", strlen("00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002d3e383cf2ca36482707617daf4230f2261cff2abeb98a7d1e139cf386970f7a67cea4e0dcf8c437e5cd9852d95613a255ef625412a3ac7fb1a0d27227a32a7c1292f14b7c189f033c91217f02b34c7832958afc7ae3bb498b29ca08277dc60d1c53bb5f07280c16238a7f99c059cbbdbbc933bef4b74d604721a09b526aac1751a4bdf0df2d303418e7e5642ac4aacc730625514c87a4bcce5369cc4c1e1d2a1ee9125e09db763e7d99fa857928fabeb94ba822d5cf1cc8f5be372683ee7089082c0ca302a243f0124cc25319d069e0c689f03e4cb32e266fffd4b8c9a5e1cb2c708dc7960531ecea4331e376d7f6604228fc0606a08bda95ee3350c8bca83f37b23160af7bae3db95f0c66ed4535fc5397b43dcdc1d09c1e3a0376a6705d916d96cb64feb47d00ebf1ddbad7eaf3b5d8c381d31098c5c8a909793bd6063c2f0450320af78de387938261eba3e984271f31c3f71a55b33631b90505f8209b384aa55feb1c1c72a5e2abce15f24eb18715a309f5517ac3079c64c8ff157d3e35d5bad17b86f9599b1e34f1f4b7c6600a83913261645a0811fba0ad1ed104fe0c") / 2);
  //   for (size_t i = 0; i < strlen("00000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000002d3e383cf2ca36482707617daf4230f2261cff2abeb98a7d1e139cf386970f7a67cea4e0dcf8c437e5cd9852d95613a255ef625412a3ac7fb1a0d27227a32a7c1292f14b7c189f033c91217f02b34c7832958afc7ae3bb498b29ca08277dc60d1c53bb5f07280c16238a7f99c059cbbdbbc933bef4b74d604721a09b526aac1751a4bdf0df2d303418e7e5642ac4aacc730625514c87a4bcce5369cc4c1e1d2a1ee9125e09db763e7d99fa857928fabeb94ba822d5cf1cc8f5be372683ee7089082c0ca302a243f0124cc25319d069e0c689f03e4cb32e266fffd4b8c9a5e1cb2c708dc7960531ecea4331e376d7f6604228fc0606a08bda95ee3350c8bca83f37b23160af7bae3db95f0c66ed4535fc5397b43dcdc1d09c1e3a0376a6705d916d96cb64feb47d00ebf1ddbad7eaf3b5d8c381d31098c5c8a909793bd6063c2f0450320af78de387938261eba3e984271f31c3f71a55b33631b90505f8209b384aa55feb1c1c72a5e2abce15f24eb18715a309f5517ac3079c64c8ff157d3e35d5bad17b86f9599b1e34f1f4b7c6600a83913261645a0811fba0ad1ed104fe0c") / 2; i++) {
  //       printf("%02X ", proof_bytes.raw[i]);
  //       if ((i + 1) % 16 == 0) {
  //           printf("\n"); // Add a newline for readability after every 16 bytes
  //       }
  //   }
  //   printf("\n");
  ctt_eth_verkle_ipa_proof_aff proof;

  ctt_eth_verkle_ipa_status stat = ctt_eth_verkle_ipa_deserialize_aff(&proof, &proof_bytes);
  printf("ctt_eth_verkle_ipa_deserialize_aff: %d\n", stat);
//   printf("proof.g2_proof.l[0].x.limbs[0]: %lu\n", proof.g2_proof.l[0].x.limbs[0]);
  
  return 0;
}
